### PR TITLE
metricbeat/module/postgresql: Add PostgreSQL Replication metricset

### LINKED ISSUE #11234: Add metricset for replication to Metricbeat Postgresql module
**Describe the enhancement:**

Add a metricset for monitoring Postgresql replication information from the `pg_stat_replication` table. There are already modules that use similar tables to extract similar data. 

https://www.postgresql.org/docs/9.5/monitoring-stats.html#PG-STAT-REPLICATION-VIEW

**Describe a specific use case for the enhancement or feature:**

This would be great for displaying and tracking the replication time between servers, time taken in each state, and the overall status of the replica set. 

Thanks guys!
-Noah

### PR DESCRIPTION
**Pull request by zxcSora**
- Enhancement
## What does this PR do?
Add metricset for replication to Metricbeat PostgreSQL module.

## Why is it important?
It is important monitoring PostgreSQL replication

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## How to test this PR locally

1) You need 2 PostgreSQL servers, Primary<->Replica. 
2) Build metricbeat from this PR
3) Run metricbeat with:
```
- module: postgresql
  metricsets:
    - replication
  hosts: ["postgres://user:password@localhost:5432/postgres"]
```

## Related issues

- Closes #11234 

## Screenshots

Screenshots from real primary production server with 2 replicas.
image
image

### CODE REVIEW

**Comment by ishleenk17** [comment]
Thanks for the contribution @zxcSora . 
We will be sharing our reviews on this soon.

**Review by shmsr — CHANGES_REQUESTED** [review]
Thank you for your contribution.

Left some comments that can easily be fixed. I'll next do a thorough technical review as well, please give me more time to review the same.

I left some comments for parts which are even autogenerated. I think we should file a separate issue to fix the template used for autogenerating these. 

Here's one: [metricset.go.tmpl](https://github.com/elastic/beats/blob/main/metricbeat/scripts/module/metricset/metricset.go.tmpl)  where some doc comments for some methods (and functions) can be improved. For example: "Fetch method implements" instead of "Fetch methods implements"

**Review by shmsr — CHANGES_REQUESTED** [review]
Thank you for your contribution.

Left some comments that can easily be fixed. I'll next do a thorough technical review as well, please give me more time to review the same.

I left some comments for parts which are even autogenerated. I think we should file a separate issue to fix the template used for autogenerating these. 

Here's one: [metricset.go.tmpl](https://github.com/elastic/beats/blob/main/metricbeat/scripts/module/metricset/metricset.go.tmpl)  where some doc comments for some methods (and functions) can be improved. For example: "Fetch method implements" instead of "Fetch methods implements"

**Comment by zxcSora** [comment]
> Thank you for your contribution.
> 
> Left some comments that can easily be fixed. I'll next do a thorough technical review as well, please give me more time to review the same.
> 
> I left some comments for parts which are even autogenerated. I think we should file a separate issue to fix the template used for autogenerating these.
> 
> Here's one: [metricset.go.tmpl](https://github.com/elastic/beats/blob/main/metricbeat/scripts/module/metricset/metricset.go.tmpl) where some doc comments for some methods (and functions) can be improved. For example: "Fetch method implements" instead of "Fetch methods implements"

Thanks for your comments. I commited changes. Waiting for technical review

File: metricbeat/module/postgresql/replication/data.go
Comment (ishleenk17) [inline metricbeat/module/postgresql/replication/data.go:41]: The backend_xmin should not be of type Time. This is 32 bit quantile, so integer perhaps ?

File: metricbeat/module/postgresql/replication/data.go
Comment (ishleenk17) [inline metricbeat/module/postgresql/replication/data.go:47]: All the lag related fields have postgresql datatype of " Interval" , which requires 16 bytes storage. 
Here we are storing it as string. Doesn't look to be appropriate ?

File: metricbeat/module/postgresql/replication/data.go
Comment (shmsr) [inline metricbeat/module/postgresql/replication/data.go:52]: Can you please double-check? For some, you have used the correct column name or for some not as an argument to c.{Str,Int,Time}

https://www.postgresql.org/docs/9.5/monitoring-stats.html#PG-STAT-REPLICATION-VIEW

File: metricbeat/module/postgresql/replication/data.go
Comment (shmsr) [inline metricbeat/module/postgresql/replication/data.go:41]: Are you sure RFC3339Nano should be used? Ref?

File: metricbeat/module/postgresql/replication/replication.go
Comment (shmsr) [inline metricbeat/module/postgresql/replication/replication.go:61]:
```suggestion
		return fmt.Errorf("error in QueryStats %w", err)
```

File: metricbeat/module/postgresql/replication/data.go
Comment (shmsr) [inline metricbeat/module/postgresql/replication/data.go:27]: Are you sure this is the correct link that wanted to put here?

File: metricbeat/module/postgresql/replication/_meta/fields.yml
Comment (ishleenk17) [inline metricbeat/module/postgresql/replication/_meta/fields.yml:5]: Did you mean "pg_stat_replication" here ?

File: metricbeat/module/postgresql/replication/_meta/fields.yml
Comment (shmsr) [inline metricbeat/module/postgresql/replication/_meta/fields.yml:74]: `application_name` is of type long? Also, the description is not correct. Please vet all names, type and description once again.

Isn't application_name is the name of the application that is connected to this WAL sender? It's also of type string.

**Review by shmsr — CHANGES_REQUESTED** [review]

File: metricbeat/module/postgresql/replication/_meta/fields.yml
Comment (ishleenk17) [inline metricbeat/module/postgresql/replication/_meta/fields.yml:55]: This should be a keyword type ?

File: metricbeat/module/postgresql/replication/_meta/fields.yml
Comment (ishleenk17) [inline metricbeat/module/postgresql/replication/_meta/fields.yml:65]: This should be a keyword type ?

File: metricbeat/module/postgresql/replication/_meta/fields.yml
Comment (ishleenk17) [inline metricbeat/module/postgresql/replication/_meta/fields.yml:71]: Keyword type ?

File: metricbeat/module/postgresql/replication/data.go
Comment (shmsr) [reply metricbeat/module/postgresql/replication/data.go:41]: Ref:
- https://www.postgresql.org/docs/current/datatype-datetime.html
- https://www.postgresql.org/docs/current/datatype-datetime.html#:~:text=tomorrow%27%3A%3Adate.-,8.5.2.%C2%A0Date/Time%20Output,-The%20output%20format

File: metricbeat/module/postgresql/replication/_meta/fields.yml
Comment (ishleenk17) [inline metricbeat/module/postgresql/replication/_meta/fields.yml:6]: Since its a new metricset, this should be beta

File: metricbeat/module/postgresql/replication/data.go
Comment (shmsr) [reply metricbeat/module/postgresql/replication/data.go:41]: ++ I also think so. From the definition, it seems it's a transaction ID.

Based on the above and [this](https://github.com/cube-js/cube/blob/master/rust/cubesql/cubesql/src/compile/engine/information_schema/postgres/pg_stat_activity.rs#L37) I think it should ideally use unsigned int.

@ishleenk17 It seems in `mapstr` package i.e., `c.Uint` isn't there and just `c.Int` is there that's using `toInt` function.

- https://pkg.go.dev/strconv#ParseUint
- https://github.com/elastic/beats/blob/main/libbeat/common/schema/mapstrstr/mapstrstr.go#L112
- https://github.com/elastic/beats/blob/main/libbeat/common/schema/mapstrstr/mapstrstr.go#L118

@ishleenk17 What do you think we should do in this case? Should we add a new function `Uint` in `mapstr` package? If yes, should we have it in the same PR or a separate PR for it?

File: metricbeat/module/postgresql/replication/data.go
Comment (zxcSora) [reply metricbeat/module/postgresql/replication/data.go:47]: > 

How do you suggest storing?

**Review by shmsr — CHANGES_REQUESTED** [review]

File: metricbeat/module/postgresql/replication/data.go
Comment (shmsr) [reply metricbeat/module/postgresql/replication/data.go]: I was just confirming. You changed it to RFC3339 which is not correct IMO. Postgres has precision up to 1 microsecond (if not nanosecond). RFC3339 Nano is fine is the datetime format is RFC3339. The recent change you made is not correct.

Comment (shmsr) [reply metricbeat/module/postgresql/replication/data.go]: @zxcSora You used `c.Int`. You might have added new functions to mapstr package if we are going to use uint32. If the correct data type is not used, values might overflow and result in incorrect metrics.

File: metricbeat/module/postgresql/replication/data.go
Comment (zxcSora) [reply metricbeat/module/postgresql/replication/data.go:47]: @shmsr Maybe you can suggest?

Comment (shmsr) [reply metricbeat/module/postgresql/replication/data.go:47]: Either we have to use a custom data type or maybe string only. Depends.

Example of a custom type: https://github.com/sanyokbig/pqinterval/blob/v1.1.2/interval.go#L11

https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT

I'd suggest doing some more research before finalizing what to use. @zxcSora please let us know what are you going to use and why.

Comment (zxcSora) [reply metricbeat/module/postgresql/replication/data.go:47]: in my opinion, sent_lsn, write_lsn, flush_lsn, replay_lsn shoud be a string, because value of this parameters it's like: 
```
sent_lsn         | 3D0/F7574390
write_lsn        | 3D0/F7574390
flush_lsn        | 3D0/F7574390
replay_lsn       | 3D0/F7574390
```
but for flush_lag, write_lag, replay_lag we need to use ```c.Time``` with HH:MM:SS.NNNNNN

**Comment by shmsr** [comment]
@zxcSora Changes look good to me but can you please update the screenshot with the latest image in the PR description?

**Comment by shmsr** [comment]
@zxcSora Can you please also post the results for the multi-host setup from where you collect metrics from multiple Postgres hosts linked by replication strategy?

File: metricbeat/module/postgresql/replication/data.go
Comment (ishleenk17) [reply metricbeat/module/postgresql/replication/data.go]: For lsn fields, string datatype looks fine.

**Comment by zxcSora** [comment]
> @zxcSora Changes look good to me but can you please update the screenshot with the latest image in the PR description?

Updated screenshots from primary server with 2 replicas

**Comment by zxcSora** [comment]
> @zxcSora Can you please also post the results for the multi-host setup from where you collect metrics from multiple Postgres hosts linked by replication strategy?
In PR description

File: metricbeat/module/postgresql/replication/replication_integration_test.go
Comment (zxcSora) [reply metricbeat/module/postgresql/replication/replication_integration_test.go:36]: I don't get you. Can you explain your question ?

Comment (shmsr) [reply metricbeat/module/postgresql/replication/replication_integration_test.go:36]: Can we write the tests with multi-hosts in mind? Like a primary server with 2 replicas? Because then only replication metrics would make sense.

**Comment by shmsr** [comment]
> > @zxcSora Changes look good to me but can you please update the screenshot with the latest image in the PR description?
> 
> Updated screenshots from primary server with 2 replicas

Thanks!

File: metricbeat/module/postgresql/replication/replication_integration_test.go
Comment (zxcSora) [reply metricbeat/module/postgresql/replication/replication_integration_test.go:36]: i think we can't do tests like this because creating cluster from docker-compose its bad idea (too much bad code)

File: metricbeat/module/postgresql/replication/replication.go
Comment (shmsr) [inline metricbeat/module/postgresql/replication/replication.go:61]:
```suggestion
		return fmt.Errorf("error in QueryStats: %w", err)
```

File: metricbeat/module/postgresql/replication/replication_integration_test.go
Comment (shmsr) [inline metricbeat/module/postgresql/replication/replication_integration_test.go:41]:
```suggestion
		t.Fatalf("Expected 0 error, had %d: %v", len(errs), errs)
```

Newline is auto in case of tests.

**Review by shmsr — CHANGES_REQUESTED** [review]
A few more nitpicks and after that I'll approve. Thanks!

**Comment by ishleenk17** [comment]
I am assuming the sample events are from the tests run on actual replica setup.
I see that lags are always 0. Does that signify that the primary and secondary are always in sync without any lag ?

File: metricbeat/module/postgresql/replication/replication.go
Comment (ishleenk17) [inline metricbeat/module/postgresql/replication/replication.go:59]: This query will now extract the lags in epoch time with datatype as int?
And then you are doing a  .unit conversion using the new conversion function in data.go?

Am, I reading it right ?

**Comment by zxcSora** [comment]
> @zxcSora I am assuming the sample events are from the tests run on actual replica setup. I see that lags are always 0. Does that signify that the primary and secondary are always in sync without any lag ?

yes, exactly

File: metricbeat/module/postgresql/replication/replication.go
Comment (zxcSora) [reply metricbeat/module/postgresql/replication/replication.go]: Fixed

File: metricbeat/module/postgresql/replication/replication_integration_test.go
Comment (zxcSora) [reply metricbeat/module/postgresql/replication/replication_integration_test.go]: Fixed

**Comment by zxcSora** [comment]
> @zxcSora Please add the changelog entries to this file. https://github.com/elastic/beats/blob/main/CHANGELOG.next.asciidoc

hope, i added it correctly

**Review by shmsr — APPROVED** [review]

File: CHANGELOG.next.asciidoc
Comment (shmsr) [inline CHANGELOG.next.asciidoc:226]:
```suggestion
- Add support for PostgreSQL replication metrics {pull}35562[35562]
```

**Review by ishleenk17 — APPROVED** [review]
Thanks for addressing multiple review comments. 
Looks good!

**Comment by zxcSora** [comment]
> Thanks for addressing multiple review comments. Looks good!

How i can request review from another codeowners ? or it will happen automatically ?

**Comment by ishleenk17** [comment]
@elastic/elastic-agent-data-plane : Could you please review this change.

File: metricbeat/module/postgresql/replication/replication_integration_test.go
Comment (fearful-symmetry) [inline metricbeat/module/postgresql/replication/replication_integration_test.go:93]: Was this meant to be `"metricsets": []string{"replication"},` instead? In case that was a copy-paste error, I ran it with the change, still got an error from `TestFetch`

**Review by fearful-symmetry — CHANGES_REQUESTED** [review]

File: metricbeat/module/postgresql/replication/replication_integration_test.go
Comment (zxcSora) [reply metricbeat/module/postgresql/replication/replication_integration_test.go:93]: i think we can't do tests like this because creating cluster from docker-compose its bad idea (too much bad code)

Comment (fearful-symmetry) [reply metricbeat/module/postgresql/replication/replication_integration_test.go:93]: Hmm, okay. Should we just remove this test, then? Find some smaller unit tests we can add here?

Comment (zxcSora) [reply metricbeat/module/postgresql/replication/replication_integration_test.go:93]: i have no idea what i can find in this situation, do you ?

Comment (fearful-symmetry) [reply metricbeat/module/postgresql/replication/replication_integration_test.go:93]: The idea solution here would be to fix the tests or fix the code.

**Comment by pierrehilbert** [comment]
@zxcSora Did you get time to review @fearful-symmetry's comments?

File: metricbeat/module/postgresql/replication/replication_integration_test.go
Comment (zxcSora) [reply metricbeat/module/postgresql/replication/replication_integration_test.go]: changed

**Comment by zxcSora** [comment]
> @zxcSora Did you get time to review @fearful-symmetry's comments?

Have no idea...

**Comment by fearful-symmetry** [comment]
So, it looks like the integration test still isn't passing:

**Comment by pierrehilbert** [comment]
@zxcSora any news?

**Comment by pierrehilbert** [comment]
I'm closing for now as we don't have any news.

**Pull request closed without merging**
